lib/networks/network_context.go

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2020 - present Kurtosis Technologies LLC.
+ * All Rights Reserved.
+ */
+
+package networks
+
+import (
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/kurtosis-tech/kurtosis-go/lib/kurtosis_service"
+	"github.com/kurtosis-tech/kurtosis-go/lib/services"
+	"github.com/palantir/stacktrace"
+	"github.com/sirupsen/logrus"
+	"os"
+	"path/filepath"
+)
+
+const (
+	ipPlaceholder = "KURTOSISSERVICEIP"
+)
+
+/*
+The identifier used for services with the network.
+*/
+type ServiceID string
+
+/*
+A package object containing the details that the NetworkContext is tracking about a node.
+*/
+type NetworkNode struct {
+	// The user-defined interface for interacting with the node.
+	// NOTE: this will need to be casted to the appropriate interface becaus Go doesn't yet have generics!
+	Service services.Service
+
+	// The Docker container ID running a given service
+	ContainerID string
+}
+
+
+type NetworkContext struct {
+	kurtosisService kurtosis_service.KurtosisService
+
+	// The dirpath ON THE SUITE CONTAINER where the suite execution volume is mounted
+	suiteExecutionVolumeDirpath string
+
+	// Filepath to the services directory, RELATIVE to the root of the suite execution volume root
+	servicesRelativeDirpath string
+
+	services map[ServiceID]NetworkNode
+}
+
+
+/*
+Creates a new NetworkContext object with the given parameters.
+
+Args:
+	kurtosisService: The Docker manager that will be used for manipulating the Docker engine during test network modification.
+	suiteExecutionVolumeDirpath: The path ON THE TEST SUITE CONTAINER where the suite execution volume is mounted
+	servicesRelativeDirpath: The dirpath where directories for each new service will be created to store file IO, which
+		is RELATIVE to the root of the suite execution volume!
+*/
+func NewNetworkContext(
+		kurtosisService kurtosis_service.KurtosisService,
+		suiteExecutionVolumeDirpath string,
+		servicesRelativeDirpath string) *NetworkContext {
+	return &NetworkContext{
+		kurtosisService: kurtosisService,
+		suiteExecutionVolumeDirpath: suiteExecutionVolumeDirpath,
+		servicesRelativeDirpath: servicesRelativeDirpath,
+		services: map[ServiceID]NetworkNode{},
+	}
+}
+
+// Gets the number of nodes in the network
+func (networkCtx *NetworkContext) GetSize() int {
+	return len(networkCtx.services)
+}
+
+/*
+Adds a service to the network with the given service ID, created using the given configuration ID.
+
+Args:
+	serviceId: The service ID that will be used to identify this node in the network.
+	initializer: The Docker container initializer that contains the logic for starting the service
+
+Return:
+	The new service
+	An availability checker which can be used to wait until the service is available, if desired
+*/
+func (networkCtx *NetworkContext) AddService(
```

Comment:
All the logic in here is mostly just copy-pasted from the old `ServiceInitializer.CreateService`